### PR TITLE
Fix asset deferring

### LINKED ISSUE #3640: Scope Hoisting for duplicate definition fails (sometimes)
# 🐛 bug report

When including the `@sentry/browser `dependency and calling it's init function the build fails.
Maybe related to #3437 

## 🎛 Configuration (.babelrc, package.json, cli command)

No config files at all, just the default. See the example below, I am running it from a `tsx` file.

## 🤔 Expected Behavior

It should build the bundle without error and should include all the code that is necessary to run sentry.

## 😯 Current Behavior

Failing with

```
🚨 Got unexpected null
  at nullthrows (.../parcel-bundler/node_modules/nullthrows/nullthrows.js:7:15)
  at BundleGraph.resolveSymbol (.../parcel-bundler/packages/core/core/src/BundleGraph.js:504:29)
  at BundleGraph.resolveSymbol (.../parcel-bundler/packages/core/core/src/public/BundleGraph.js:128:27)
  at resolveSymbol (.../parcel-bundler/packages/shared/scope-hoisting/src/concat.js:184:32)
  at markUsed (.../parcel-bundler/packages/shared/scope-hoisting/src/concat.js:178:9)
  at enter (.../parcel-bundler/packages/core/core/src/Graph.js:531:16)
  at enter (.../parcel-bundler/packages/core/core/src/Graph.js:531:16)
  at enter (.../parcel-bundler/packages/core/core/src/Graph.js:531:16)
  at enter (.../parcel-bundler/packages/core/core/src/Graph.js:372:26)
  at walk (.../parcel-bundler/packages/core/core/src/Graph.js:392:22)
```

## 💁 Possible Solution

When logging the dependency in `BundleGraph.js:504` it is failing to resolve `./baseBackend.js` (sometimes also somewhere else, e.g. `./api.js`):
```
{
  moduleSpecifier: './basebackend',
  loc: { line: 5, column: 28 },
  env: {
    context: 'browser',
    engines: { browsers: [Array] },
    includeNodeModules: true,
    outputFormat: 'global',
    isLibrary: false
  },
  sourceAssetId: '7738236bb0d2b80bd3a62bc913f03b89',
  sourcePath: '.../project/node_modules/@sentry/core/esm/index.js',
  id: 'b25413759777ee0e1c0de158dda1eb15',
  meta: {},
  symbols: Map {
    'BaseBackend' => '$bb0d2b80bd3a62bc913f03b89$import$BaseBackend'
  },
  isWeak: true
}
```

In the first place, I thought it's due to the `BaseClient` which is declared as `var` and `function` within different scopes. This pattern is the same in all files that I saw failing, and after renaming by hand it went to the next function, it looks like:

``` js
const Test = (function() {
  function Test() {
    console.log('init');
  }
  Test.prototype.doStuff = function() {
    console.log('do stuff');
  };
  return Test;
})();

export { Test };
```

On the other hand, the bug does not happen all the time. In most cases, it is failing, but sometimes, when deleting the cache before, it even succeeds. Therefore, it looks like a race condition to me. E.g. I run the command twice after another, the first succeeds the second failed, without any code changes:

``` bash
$ rm -rf .parcel-cache && ../../npm-packages/parcel-original/packages/core/parcel/src/bin.js build src/index.tsx
✨ Built in 24.75s.
dist/index.js                                                             54.27 KB    2.38s
├── node_modules/@sentry/browser/esm/integrations/breadcrumbs.js          17.32 KB    854ms
├── node_modules/@sentry/hub/esm/hub.js                                   15.56 KB    246ms
├── node_modules/@sentry/core/esm/baseclient.js                           13.26 KB    465ms
├── node_modules/tslib/tslib.es6.js                                       13.03 KB    284ms
├── node_modules/@sentry/utils/esm/object.js                              11.36 KB    474ms
├── node_modules/@sentry/hub/esm/scope.js                                 11.01 KB    159ms
├── node_modules/@sentry/utils/esm/misc.js                                 9.88 KB    3.24s
├── node_modules/@sentry/browser/esm/helpers.js                            9.25 KB    224ms
├── ../../npm-packages/parcel-original/node_modules/process/browser.js     8.83 KB     82ms
└── node_modules/@sentry/browser/esm/tracekit.js                            7.6 KB    170ms
└── + 45 more assets

$ rm -rf .parcel-cache && ../../npm-packages/parcel-original/packages/core/parcel/src/bin.js build src/index.tsx
🚨 Got unexpected null
  at nullthrows (.../parcel-original/node_modules/nullthrows/nullthrows.js:7:15)
  at BundleGraph.resolveSymbol (.../parcel-original/packages/core/core/src/BundleGraph.js:504:29)
  at BundleGraph.resolveSymbol (.../parcel-original/packages/core/core/src/public/BundleGraph.js:128:27)
  at resolveSymbol (.../parcel-original/packages/shared/scope-hoisting/src/concat.js:184:32)
  at markUsed (.../parcel-original/packages/shared/scope-hoisting/src/concat.js:178:9)
  at enter (.../parcel-original/packages/core/core/src/Graph.js:531:16)
  at enter (.../parcel-original/packages/core/core/src/Graph.js:531:16)
  at enter (.../parcel-original/packages/core/core/src/Graph.js:531:16)
  at enter (.../parcel-original/packages/core/core/src/Graph.js:372:26)
  at walk (.../parcel-original/packages/core/core/src/Graph.js:392:22)
```

## 💻 Code Sample

Using `@sentry/browser@5.7.0`

``` js
import * as Sentry from '@sentry/browser';

Sentry.init({
  dsn: '...',
});
```

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | v2 (commit eedc965377940034cf78a026d581f12098df724e)
| Node             | v12.8.1 
| npm         | v6.11.2
| Operating System | Windows (Ubuntu Bash)

### PR DESCRIPTION
**Pull request by mischnic**
Fixes #3640

(Not sure if there's something that can be optimized for performance?)

## Problem description

Deferred dependencies weren't rechecked if another dependency to an already added asset was added (e.g. importing another function from lodash from a different asset).

66719659-2ed41e80-edf3-11e9-8102-905cbfa90191

## Solution

1. Move the deferring logic to asset graph builder (so that an asset group can be added the request graph when needed). 

https://github.com/parcel-bundler/parcel/blob/000954b5d81268784ae201e5d00a0f76004f4896/packages/core/core/src/AssetGraphBuilder.js#L256-L260

2. The dependencies (that are already resolved) are not discarded when a dependency should be deferred but are instead added with `deferred = true`. This way, only these are rechecked in the situation above.

AssetGraph

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>packages/benchmarks/kitchen-sink</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>packages/benchmarks/react-hn</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 15.53s | -69.00ms |
| Cached | 9.72s | +96.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/main/index.js | 34.53kb | +0.00b | 1.74s | -728.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/main/index.js | 34.53kb | +0.00b | 12.00ms | -6.00ms 🚀 |
| dist/module/index.js | 34.33kb | +0.00b | 9.00ms | -5.00ms 🚀 |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/247235588769448462)**

**Comment by garthenweb** [comment]
@mischnic Great work! I just checked this branch against the two projects where I experienced this issue and they are both compiling now :) Thank you!

File: packages/core/core/src/AssetGraphBuilder.js [hunk 1]
```diff
@@ -164,11 +171,105 @@ export default class AssetGraphBuilder extends EventEmitter {
...
+    if (existingAssetGroupNode) {
+      // Don't overwrite non-deferred asset groups with deferred ones
+      invariant(existingAssetGroupNode.type === 'asset_group');
+      assetGroupNode.deferred = existingAssetGroupNode.deferred && defer;
+    }
+    this.assetGraph.resolveDependency(dependency, assetGroupNode);
+    if (existingAssetGroupNode) {
+      // Node already existed, that asset might have deferred dependencies,
+      // recheck all dependencies of all assets of this asset group
+      let assetNodes = this.assetGraph
```
Comment (devongovett) [inline packages/core/core/src/AssetGraphBuilder.js]: Maybe use `traverse` for this? You can use `skipChildren` to stop going any further once you hit the asset group.

Comment (mischnic) [reply packages/core/core/src/AssetGraphBuilder.js]: Great suggestion, that section is way more concise now.

**Comment by mischnic** [comment]
Ready for review, the added tests are now deterministic thanks to @padmaia's suggestion.

**Pull request merged**